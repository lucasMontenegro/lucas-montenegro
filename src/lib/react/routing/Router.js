import Translation from "lib/Translation"
import languageDetector from "lib/languageDetector"
export default class Router {
  constructor (routing) {
    this.matchers = routing.matchers
    const initialLanguage = routing.languageCodes[0]
    this.locations = [`home`, `notFound`].reduce((byClient, clientName) => {
      const location = routing.locations[clientName]
      const translators = routing.linkTranslators[clientName]
      const intl = translators[initialLanguage].toIntl(location)
      byClient[clientName] = new Translation(
        routing.languageCodes.reduce((byLanguage, newLanguage) => {
          byLanguage[newLanguage] = (
            initialLanguage === newLanguage ? location : translators[newLanguage].toLocal(intl)
          )
          return byLanguage
        }, {})
      )
      return byClient
    }, {})
  }
  findRoute (location) {
    // detect root route
    if (this.matchers.root(location)) {
      return {
        render: {},
        redirect: this.locations.home.get(),
      }
    }
    { // detect language-only route
      const matcher = this.matchers.languageOnly.find(m => m.match(location))
      if (matcher) {
        languageDetector.set(matcher.languageCode)
        return {
          render: {},
          redirect: this.locations.home.get(),
        }
      }
    }
    { // find client route
      const matcher = this.matchers.client.find(m => m.match(location))
      if (matcher) {
        languageDetector.set(matcher.languageCode)
        const { clientName } = matcher
        return {
          clientName,
          render: { [clientName]: true },
          redirect: null,
        }
      }
    }
    { // detect unknown client
      const matcher = this.matchers.unknownClient.find(m => m.match(location))
      if (matcher) {
        languageDetector.set(matcher.languageCode)
        return {
          render: {},
          redirect: {
            ...this.locations.notFound.get(),
            state: location,
          },
        }
      }
    }
    return { // 404 redirection
      render: {},
      redirect: {
        ...this.locations.notFound.get(),
        state: location,
      },
    }
  }
}