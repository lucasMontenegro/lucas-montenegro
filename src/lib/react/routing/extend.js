import Translation from "lib/Translation"
export default function extendRouting (routing) {
  const initialLanguage = routing.languageCodes[0]
  routing.translatedLocations = Object.keys(routing.locations).reduce((byClient, clientName) => {
    const location = routing.locations[clientName]
    const translators = routing.linkTranslators[clientName]
    const intl = translators[initialLanguage].toIntl(location)
    byClient[clientName] = new Translation(
      routing.languageCodes.reduce((byLanguage, newLanguage) => {
        byLanguage[newLanguage] = (
          initialLanguage === newLanguage ? location : translators[newLanguage].toLocal(intl)
        )
        return byLanguage
      }, {})
    )
    return byClient
  }, {})
  return routing
}