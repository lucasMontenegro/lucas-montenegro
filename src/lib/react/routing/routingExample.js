export default {
  clientNames: [`home`, `foo`, `notFound`],
  languageCodes: [`en`, `es`],
  languageNames: {
    en: `English`,
    es: `Spanish`,
  },
  locations: {
    home: { pathname: `/react/routing/en/home/2` },
    foo: { pathname: `/react/routing/en/foo/5` },
    notFound: { pathname: `/react/routing/en/notFound` },
  },
  matchers: {
    root (location) {
      return /^\/react\/routing\/?$/.test(location.pathname)
    },
    languageOnly: [
      {
        languageCode: `en`,
        match (location) {
          return /^\/react\/routing\/en\/?$/.test(location.pathname)
        },
      },
      {
        languageCode: `es`,
        match (location) {
          return /^\/react\/routing\/es\/?$/.test(location.pathname)
        },
      },
    ],
    client: [
      {
        clientName: `home`,
        languageCode: `en`,
        match (location) {
          return /^\/react\/routing\/en\/home\/\d\/?$/.test(location.pathname)
        },
      },
      {
        clientName: `home`,
        languageCode: `es`,
        match (location) {
          return /^\/react\/routing\/es\/home\/\d\/?$/.test(location.pathname)
        },
      },
      {
        clientName: `foo`,
        languageCode: `en`,
        match (location) {
          return /^\/react\/routing\/en\/foo\/\d\/?$/.test(location.pathname)
        },
      },
      {
        clientName: `foo`,
        languageCode: `es`,
        match (location) {
          return /^\/react\/routing\/es\/foo\/\d\/?$/.test(location.pathname)
        },
      },
      {
        clientName: `notFound`,
        languageCode: `en`,
        match (location) {
          return /^\/react\/routing\/en\/notFound\/?$/.test(location.pathname)
        },
      },
      {
        clientName: `notFound`,
        languageCode: `es`,
        match (location) {
          return /^\/react\/routing\/es\/notFound\/?$/.test(location.pathname)
        },
      },
    ],
    unknownClient: [
      {
        languageCode: `en`,
        match (location) {
          return /^\/react\/routing\/en\/.+/.test(location.pathname)
        },
      },
      {
        languageCode: `es`,
        match (location) {
          return /^\/react\/routing\/es\/.+/.test(location.pathname)
        },
      },
    ],
  },
  linkTranslators: {
    home: {
      en: {
        toIntl(location) {
          return location.pathname.charAt(23)
        },
        toLocal(n) {
          return { pathname: `/react/routing/en/home/${n}` }
        },
      },
      es: {
        toIntl(location) {
          return location.pathname.charAt(23)
        },
        toLocal(n) {
          return { pathname: `/react/routing/es/home/${n}` }
        },
      },
    },
    foo: {
      en: {
        toIntl(location) {
          return location.pathname.charAt(22)
        },
        toLocal(n) {
          return { pathname: `/react/routing/en/foo/${n}` }
        },
      },
      es: {
        toIntl(location) {
          return location.pathname.charAt(22)
        },
        toLocal(n) {
          return { pathname: `/react/routing/es/foo/${n}` }
        },
      },
    },
    notFound: {
      en: {
        toIntl(location) {
          return location.state
        },
        toLocal(state) {
          return { pathname: `/react/routing/en/notFound`, state }
        },
      },
      es: {
        toIntl(location) {
          return location.state
        },
        toLocal(state) {
          return { pathname: `/react/routing/es/notFound`, state }
        },
      },
    },
  },
}