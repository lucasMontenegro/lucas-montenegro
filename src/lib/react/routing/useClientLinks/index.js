import useBaseLinks from "./useBaseLinks"
import useClientLocations from "./useClientLocations"
import { useMemo } from "react"
export default function useClientLinks (render, routing, route) {
  const baseLinks = useBaseLinks(render)
  const locations = useClientLocations(routing, route)
  return useMemo(() => {
    if (locations === null) {
      return []
    }
    const currentClient = route.clientName
    return baseLinks.map(link => {
      const { clientName } = link
      return {
        clientName,
        active: currentClient === clientName,
        render: link.render,
        location: locations[clientName],
      }
    })
  }, [locations, baseLinks, route])
}