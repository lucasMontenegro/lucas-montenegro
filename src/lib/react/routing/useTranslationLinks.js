import { useMemo } from "react"
import languageDetector from "lib/languageDetector"
export default function useTranslationLinks (routing, route) {
  const baseLinks = useMemo(() => {
    const { languageNames } = routing
    return routing.languageCodes.map(languageCode => ({
      languageCode,
      text: languageNames[languageCode],
    }))
  }, [routing])
  const { clientName } = route
  const translators = useMemo(() => (
    clientName ? routing.linkTranslators[clientName] : null
  ), [routing, clientName])
  return useMemo(() => {
    if (translators) {
      let links
      return function getTranslationLinks () {
        if (links) {
          return links
        }
        const oldLanguage = languageDetector.get()
        const intl = translators[oldLanguage].toIntl(route.location)
        return links = baseLinks.filter(link => link.languageCode !== oldLanguage).map(link => {
          const { languageCode } = link
          return {
            languageCode,
            location: translators[languageCode].toLocal(intl),
            text: link.text,
          }
        })
      }
    }
    const links = []
    return () => links
  }, [translators, route, baseLinks])
}