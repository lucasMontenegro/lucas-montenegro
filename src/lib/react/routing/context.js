import React, { createContext, useMemo, useContext } from "react"
import Router from "./Router"
import useClientLinks from "./useClientLinks"
import useTranslationLinks from "./useTranslationLinks"
export const RoutingContext = createContext({
  routing: null,
  route: {
    render: {},
    redirect: null,
  },
})
RoutingContext.displayName = `RoutingContext`
export function RoutingProvider (props) {
  const { routing, location } = props
  const router = useMemo(() => new Router(routing), [routing])
  const route = useMemo(() => router.findRoute(location), [router, location])
  const clientLinks = useClientLinks(props.clientLinks, routing, route)
  const getTranslationLinks = useTranslationLinks(routing, route)
  const value = useMemo(() => (
    { routing, route, clientLinks, getTranslationLinks }
  ), [routing, route, clientLinks, getTranslationLinks])
  return <RoutingContext.Provider value={value}>{props.children}</RoutingContext.Provider>
}
export function useRoute () {
  return useContext(RoutingContext).route
}
export function useRoutingContext () {
  return useContext(RoutingContext)
}