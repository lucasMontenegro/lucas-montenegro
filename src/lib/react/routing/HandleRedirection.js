import { useRoute } from "./context"
import React from "react"
import { Redirect } from "react-router-dom"
export default function HandleRedirection () {
  const route = useRoute()
  if (route.redirect) {
    return <Redirect to={route.redirect} />
  }
  return null
}