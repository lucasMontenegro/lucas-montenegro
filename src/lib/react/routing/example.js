import React, { useState } from "react"
import Div from "lib/react/utils/Div"
import StringifyObject from "lib/react/utils/StringifyObject"
import languageDetector from "lib/languageDetector"
import Link from "lib/react/links/Link"
import { useRoute, useRoutingContext, RoutingProvider } from "./context"
import GetValue from "lib/react/utils/GetValue"
import HandleRedirection from "./HandleRedirection"
import routingExample from "./routingExample"
import { Route } from "react-router-dom"
function CurrentLanguage () {
  return (
    <Div color="Green">
      <h4>languageDetector.get()</h4>
      <StringifyObject id="current-language" source={languageDetector.get()} />
    </Div>
  )
}
function CurrentLocation (props) {
  return (
    <Div color="RoyalBlue">
      <h4>props.location</h4>
      <StringifyObject id="current-location" source={props.location} />
    </Div>
  )
}
function TestLinks () {
  return (
    <Div color="Orange">
      <h4>Test Links</h4>
      <ul>
        {testLinks.map(list => (
          <li key={list.title}>
            <h5>{list.title}</h5>
            <ul>
              {list.links.map(link => {
                const text = link.text || link.location.pathname
                return <li key={text}><Link to={link.location}>{text}</Link></li>
              })}
            </ul>
          </li>
        ))}
      </ul>
    </Div>
  )
}
function UseRoute () {
  const route = useRoute()
  return (
    <Div color="Purple">
      <h4>useRoute()</h4>
      <StringifyObject id="use-route" source={route} />
      {route.redirect && <div><Link to={route.redirect}>route.redirect</Link></div>}
    </Div>
  )
}
function Redirection () {
  const [enabled, saveEnabled] = useState(false)
  return (
    <Div color="Red">
      <h4>Handle Redirection</h4>
      <GetValue
        id="enable-redirection"
        value={enabled}
        onClick={() => saveEnabled(!enabled)}
      />
      {enabled ? <HandleRedirection /> : null}
    </Div>
  )
}
function ClientLinks () {
  const [count, setCount] = useState(0)
  return (
    <Div color="Yellow">
      <h4>Client Links</h4>
      <GetValue
        id="re-render-client-links"
        value={count}
        onClick={() => setCount(n => n + 1)}
      />
      <ul>
        {useRoutingContext().clientLinks.map(link => (
          <li key={link.clientName}>
            <Link to={link.location}>{link.render}</Link>
            <StringifyObject id={`${link.clientName}-client-link`} source={link} />
          </li>
        ))}
      </ul>
    </Div>
  )
}
function TranslationLinks () {
  const [links, saveLinks] = useState(null)
  const [render, setRender] = useState(true)
  const context = useRoutingContext()
  return (
    <Div color="Teal">
      <h4>Translation Links</h4>
      <GetValue
        id="render-translation-links"
        value={render}
        onClick={() => setRender(!render)}
      />
      <GetValue
        id="get-translation-links"
        value={links}
        onClick={() => saveLinks(context.translationLinks.get(render))}
      />
      <ul>
        {links && links.map(link => {
          const { languageCode } = link
          return (
            <li key={languageCode}>
              <Link to={link.location}>{link.text}</Link>
              <StringifyObject id={`${languageCode}-translation-link`} source={link} />
            </li>
          )
        })}
      </ul>
    </Div>
  )
}
function Example (props) {
  languageDetector.init(routingExample.languageCodes)
  return languageDetector.useReadyState() ? (
    <RoutingProvider
      routing={routingExample}
      location={props.location}
      clientLinks={{ home: `Home`, foo: `Foo`, notFound: `Not Found` }}
    >
      <CurrentLanguage />
      <CurrentLocation location={props.location} />
      <TestLinks />
      <UseRoute />
      <Redirection />
      <ClientLinks />
      <TranslationLinks />
    </RoutingProvider>
  ) : null
}
export default (<Route path="/react/routing" component={Example} />)
const testLinks = [
  {
    title: `Root`,
    links: [
      { location: { pathname: `/react/routing` } },
      { location: { pathname: `/react/routing/` } },
    ],
  },
  {
    title: `Language-only`,
    links: [
      { location: { pathname: `/react/routing/en` } },
      { location: { pathname: `/react/routing/en/` } },
      { location: { pathname: `/react/routing/es` } },
      { location: { pathname: `/react/routing/es/` } },
    ],
  },
  {
    title: `Home`,
    links: [
      { location: { pathname: `/react/routing/en/home/4` } },
      { location: { pathname: `/react/routing/en/home/3/` } },
      { location: { pathname: `/react/routing/es/home/4` } },
      { location: { pathname: `/react/routing/es/home/3/` } },
    ],
  },
  {
    title: `Foo`,
    links: [
      { location: { pathname: `/react/routing/en/foo/1` } },
      { location: { pathname: `/react/routing/en/foo/6/` } },
      { location: { pathname: `/react/routing/es/foo/1` } },
      { location: { pathname: `/react/routing/es/foo/6/` } },
    ],
  },
  {
    title: `Not Found`,
    links: [
      { location: { pathname: `/react/routing/en/notFound` } },
      { location: { pathname: `/react/routing/en/notFound/` } },
      {
        text: `/react/routing/en/notFound with state`,
        location: {
          pathname: `/react/routing/en/notFound`,
          state: { pathname: `/react/routing/404` },
        },
      },
      { location: { pathname: `/react/routing/es/notFound` } },
      { location: { pathname: `/react/routing/es/notFound/` } },
      {
        text: `/react/routing/es/notFound with state`,
        location: {
          pathname: `/react/routing/es/notFound`,
          state: { pathname: `/react/routing/404` },
        },
      },
    ],
  },
  {
    title: `Unknown Client`,
    links: [
      { location: { pathname: `/react/routing/en/404` } },
      { location: { pathname: `/react/routing/es/404` } },
    ],
  },
  {
    title: `No Match`,
    links: [
      { location: { pathname: `/react/routing/404` } },
      { location: { pathname: `/react/routing/pt/foo` } },
    ],
  },
]