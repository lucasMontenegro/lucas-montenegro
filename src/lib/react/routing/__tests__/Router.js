import Translation from "lib/Translation"
import languageDetector from "lib/languageDetector"
import Router from "../Router"
import routing from "../routingExample"
jest.mock(`lib/Translation`, () => ({ __esModule: true, default: jest.fn() }))
let currentLanguage
Translation.mockImplementation(obj => ({ get: () => obj[currentLanguage] }))
jest.mock(`lib/languageDetector`, () => ({ __esModule: true, default: {} }))
describe(`../Router`, () => {
  it(`should use the right versions of its dependencies`, () => {
    expect(jestUtils.getDependencies([`react-router-dom`])).toMatchSnapshot()
  })
  describe(`new Router`, () => {
    let router
    beforeAll(() => {
      router = new Router(routing)
    })
    {
      const msg = (
        `should create the initial client locations (client name %s, current language %s)`
      )
      const cases = [
        [`home`, `en`, { pathname: `/react/routing/en/home/2` }],
        [`home`, `es`, { pathname: `/react/routing/es/home/2` }],
        [`notFound`, `en`, { pathname: `/react/routing/en/notFound` }],
        [`notFound`, `es`, { pathname: `/react/routing/es/notFound` }],
      ]
      test.each(cases)(msg, (clientName, languageCode, expected) => {
        currentLanguage = languageCode
        expect(router.locations[clientName].get()).toEqual(expected)
      })
    }
    describe(`router.findRoute (root route)`, () => {
      let route
      beforeAll(() => {
        currentLanguage = `en`
        route = router.findRoute({ pathname: `/react/routing` })
      })
      it(`should return a route that redirects to home`, () => {
        expect(route).toEqual({
          render: {},
          redirect: { pathname: `/react/routing/en/home/2` },
        })
      })
    })
    describe(`router.findRoute (language-only route)`, () => {
      let route
      const setLanguage = jest.fn(str => {
        currentLanguage = str
      })
      beforeAll(() => {
        languageDetector.set = setLanguage
        currentLanguage = `es`
        route = router.findRoute({ pathname: `/react/routing/en` })
        delete languageDetector.set
      })
      it(`should return a route that redirects to home`, () => {
        expect(route).toEqual({
          render: {},
          redirect: { pathname: `/react/routing/en/home/2` },
        })
      })
      it(`should call languageDetector.set`, () => {
        expect(setLanguage.mock.calls).toEqual([[`en`]])
      })
    })
    describe(`router.findRoute (client route)`, () => {
      let route
      const setLanguage = jest.fn()
      beforeAll(() => {
        languageDetector.set = setLanguage
        currentLanguage = `en`
        route = router.findRoute({ pathname: `/react/routing/es/foo/8` })
        delete languageDetector.set
      })
      it(`should return a route that renders the proper client`, () => {
        expect(route).toEqual({
          clientName: `foo`,
          render: { foo: true },
          redirect: null,
        })
      })
      it(`should call languageDetector.set`, () => {
        expect(setLanguage.mock.calls).toEqual([[`es`]])
      })
    })
    describe(`router.findRoute (detect unknown client)`, () => {
      let route
      const setLanguage = jest.fn(str => {
        currentLanguage = str
      })
      const location = { pathname: `/react/routing/en/bar` }
      beforeAll(() => {
        languageDetector.set = setLanguage
        currentLanguage = `es`
        route = router.findRoute(location)
        delete languageDetector.set
      })
      it(`should return a route that redirects to notFound`, () => {
        expect(route).toEqual({
          render: {},
          redirect: {
            pathname: `/react/routing/en/notFound`,
            state: { pathname: `/react/routing/en/bar` },
          },
        })
        expect(route.redirect.state).toBe(location)
      })
      it(`should call languageDetector.set`, () => {
        expect(setLanguage.mock.calls).toEqual([[`en`]])
      })
    })
    describe(`router.findRoute (404 redirection)`, () => {
      let route
      const location = { pathname: `/react/routing/404` }
      beforeAll(() => {
        currentLanguage = `es`
        route = router.findRoute(location)
      })
      it(`should return a route that redirects to notFound`, () => {
        expect(route).toEqual({
          render: {},
          redirect: {
            pathname: `/react/routing/es/notFound`,
            state: { pathname: `/react/routing/404` },
          },
        })
        expect(route.redirect.state).toBe(location)
      })
    })
  })
})