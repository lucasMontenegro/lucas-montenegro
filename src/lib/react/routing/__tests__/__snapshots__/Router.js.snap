// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`../Router should use the right versions of its dependencies 1`] = `
Object {
  "react-router-dom": "^5.1.2",
}
`;
